File: approxPiMT.java
import java.util.Random;
public class approxPiMT{
	static int R = 1;
	static double insides[];
	public static void main(String args[]){
		Random r = new Random();
		
		int nPoints = Integer.parseInt(args[0]);
		int nThreads = Integer.parseInt(args[1]);
		
		Thread threads[] = new Thread[nThreads]; 		
		double totalInside = 0;
		
		insides = new double[nThreads];
		

		for(int i = 0; i<nThreads;i++){
			Thread t = step(r, nPoints, nThreads, i);
			t.start();			
			threads[i] = t;
		}
		
		try{
		for (Thread t : threads){
    			t.join();
		}
		}catch(InterruptedException e){
			e.printStackTrace();
		}	
		
		for(int i = 0; i < nThreads;i++){
			totalInside+= insides[i];
		}
			
		System.out.println("Total Number of points: " + nPoints);
		System.out.println("Points within circle: " + totalInside);
		System.out.println("Pi estimation: " + 4*(totalInside/nPoints));	
	}
	
	public static Thread step(Random r, int nPoints, int nThreads, int threadID){
		return new Thread(() -> {
			double countInside = 0;
			for(int i=0; i<nPoints/nThreads;i++){
				System.out.println(i  + " on " + threadID);
				if(isInsideCircle(r.nextDouble(), r.nextDouble()))
					countInside++;
			}
			insides[threadID] = countInside;
		});
	}
	
	
	private static boolean isInsideCircle(double x, double y){
		return x*x + y*y <= R;	
	}
}

File: approxPiMT.c
#include <stdio.h> 
#include <stdlib.h> 
#include <pthread.h> 
  
  
  //compile with my exe: a.o b.o c.o d.o	gcc -o myexe a.o b.o c.o d.o(ou $^)
  //a.o: a.c	gcc -c a.c
  
  //%.o: %.		gcc -c $^
  
  
  //name for the file is makefile
  
  //program make

int nPoints;
int nThreads;


float r(unsigned int seed)
{	
    return (double)rand_r(&seed) / (double)RAND_MAX ;
}

static int isInsideCircle(float x, float y){
	return x*x + y*y <= 1;	
}

	
//Thread function get an argument and return something - void * means any type
void *f(void *arg) 
{ 	
	
	int* countInside = malloc(sizeof(int*));
	int i;
	unsigned int seed = time(NULL);
	for(i=0; i<nPoints/nThreads;i++){
		
		seed++;
		if(isInsideCircle(r(seed), r(seed))){
			(*countInside)++;
			//printf("%d\n", *countInside);
		}
	}
	
    return (void *) countInside; 
} 


//gcc -pthread -o approxPiMT approxPiMT.c
int main(int argc, char **argv)
{
	nPoints = atoi(argv[1]);
	nThreads = atoi(argv[2]);
	
	pthread_t tid;
	
	pthread_create(&tid, NULL, f, NULL);
	
	void* partial_count_inside;
	
	pthread_join(tid, &partial_count_inside);
	
	
	int countInside = 0;
	countInside = *(int *)partial_count_inside;
	
	printf("\nTotal Number of points: %d", nPoints);
	printf("\nPoints within circle: %d", countInside);
	
	double pi = (countInside/nPoints);
	printf("\nPi estimation: %f\n\n", pi);
	
    exit(0); 
}

   

